system-tests/lib/dep-installer/index.ts

```diff
@@ -0,0 +1,298 @@
+    const hasYarnLock = lockFilename === 'yarn.lock'
+
+    // 1. Ensure there is a cache directory set up for this test project's `node_modules`.
+    await ensureCacheDir(cacheNodeModulesDir)
+
+    let persistCacheCb: () => Promise<void>
+
+    if (hasYarnLock) {
+      await symlinkNodeModulesFromCache(tmpNodeModulesDir, cacheNodeModulesDir)
+    } else {
+      // due to an issue in NPM, we cannot have `node_modules` be a symlink. fall back to copying.
+      // https://github.com/npm/npm/issues/10013
+      persistCacheCb = await copyNodeModulesFromCache(tmpNodeModulesDir, cacheNodeModulesDir)
+    }
+
+    // 2. Before running the package installer, resolve workspace deps to absolute paths.
+    // This is required to fix install for workspace-only packages.
+    const workspaceDeps = await makeWorkspacePackagesAbsolute(projectPkgJsonPath)
+
+    // 3. Delete cached workspace packages since the pkg manager will create a fresh symlink during install.
+    await removeWorkspacePackages(workspaceDeps)
+
+    // 4. Fix relative paths in temp dir's lockfile.
+    const lockFilePath = path.join(projectDir, lockFilename)
+
+    log(`Writing ${lockFilename} with fixed relative paths to temp dir`)
+    await restoreLockFileRelativePaths({ projectDir, lockFilePath, relativePathToMonorepoRoot })
+
+    // 5. Run `yarn/npm install`.
+    const getCommandFn = hasYarnLock ? getYarnCommand : getNpmCommand
+    const cmd = getCommandFn({
+      updateLockFile,
+      yarnV311: pkgJson._cyYarnV311,
+      isCI: !!process.env.CI,
+      runScripts: pkgJson._cyRunScripts,
+    })
+
+    await runCmd(cmd)
+
+    // 6. Now that the lockfile is up to date, update workspace dependency paths in the lockfile with monorepo
+    // relative paths so it can be the same for all developers
+    log(`Copying ${lockFilename} and fixing relative paths for ${project}`)
+    await normalizeLockFileRelativePaths({ project, projectDir, lockFilePath, lockFilename, relativePathToMonorepoRoot })
+
+    // 7. After install, we must now symlink *over* all workspace dependencies, or else
+    // `require` calls from installed workspace deps to peer deps will fail.
+    await removeWorkspacePackages(workspaceDeps)
+    for (const dep of workspaceDeps) {
+      const destDir = path.join(tmpNodeModulesDir, dep)
+      const targetDir = pathToPackage(dep)
+
+      log(`Symlinking workspace dependency: ${dep} (${destDir} -> ${targetDir})`)
+
+      await fs.mkdir(path.dirname(destDir), { recursive: true })
+      await fs.symlink(targetDir, destDir, 'junction')
+    }
+
+    // 8. If necessary, ensure that the `node_modules` cache is updated by copying `node_modules` back.
+    if (persistCacheCb) await persistCacheCb()
+  } catch (err) {
+    if (err.code === 'MODULE_NOT_FOUND') return
+
+    console.error(`⚠ An error occurred while installing the node_modules for ${project}.`)
+    console.error(err)
+    throw err
+  }
+}
+
+export async function scaffoldCommonNodeModules () {
```

Comment:
Can you add a comment above this line describing what it does?

Reply:
40c73d0edba98d86faf817cca6e633058e62ef23